cloud/credentials.go

```diff
@@ -437,3 +433,86 @@ func RemoveSecrets(
 	}
 	return &Credential{authType: credential.authType, attributes: redactedAttrs}, nil
 }
+
+// CredentialCollection holds CloudCredential(s) that are lazily validated.
+type CredentialCollection struct {
+	Credentials map[string]interface{} `yaml:"credentials"`
+}
+
+// ParseCredentialCollection parses YAML bytes for the credential
+func ParseCredentialCollection(data []byte) (*CredentialCollection, error) {
+	collection := CredentialCollection{}
+	err := yaml.Unmarshal(data, &collection)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot unmarshal yaml credentials")
+	}
+	return &collection, nil
+}
+
+// CloudCredential returns a copy of the CloudCredential for the specifed cloud or
+// an error when the CloudCredential was not found or failed to pass validation.
+func (c *CredentialCollection) CloudCredential(cloudName string) (*CloudCredential, error) {
+	credentialValue, ok := c.Credentials[cloudName]
+	if !ok {
+		return nil, errors.NotFoundf("credentials for cloud %s", cloudName)
+	}
+	if credential, ok := credentialValue.(CloudCredential); ok {
+		return &credential, nil
+	}
+	credentialValue, err := cloudCredentialChecker{}.Coerce(
+		credentialValue, []string{"credentials." + cloudName},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	credential := credentialValue.(CloudCredential)
+	// Validate default credential actually exists in the CloudCredential
+	if credential.DefaultCredential != "" {
+		stillHaveDefault := false
+		for name := range credential.AuthCredentials {
+			if name == credential.DefaultCredential {
+				stillHaveDefault = true
+				break
+			}
+		}
+		if !stillHaveDefault {
+			credential.DefaultCredential = ""
+		}
+	}
+	c.Credentials[cloudName] = credential
+	return &credential, nil
+}
+
+// CloudNames returns all the names of CloudCredentials inside the CredentialCollection.
+func (c *CredentialCollection) CloudNames() []string {
+	var cloudNames []string
+	for k := range c.Credentials {
+		cloudNames = append(cloudNames, k)
+	}
+	return cloudNames
+}
+
+// UpdateCloudCredential stores CloudCredential for a specific cloud.
+func (c *CredentialCollection) UpdateCloudCredential(cloudName string, details CloudCredential) {
+	if len(details.AuthCredentials) == 0 {
+		delete(c.Credentials, cloudName)
+		return
+	}
+	if c.Credentials == nil {
+		c.Credentials = make(map[string]interface{})
+	}
+	// Validate default credential actually exists in the CloudCredential
```

Comment:
This block of code could be broken out as a helper as it's also used in CloudCredential(cloudName string)